### ISSUE TITLE
GDAL_DISABLE_READDIR_ON_OPEN seems to lead to failing reads from S3

### ISSUE THREAD

**Issue by dionhaefner** [OSGeo/gdal#1960]
This is a very weird and slippery bug that has plagued us via rasterio for many months. I apologize in advance for the vague description and poor reproducibility. I am reporting this in the hope that someone else has seen similar behavior, or has an idea why this could occur.

### Background

We run a [Python tile server](https://github.com/DHI-GRAS/terracotta) on AWS Lambda. Tiles are read from cloud-optimized GTiff located in an S3 bucket. After a while, we noticed that more and more files would fail to be read in an erratic fashion. Errors looked like this in the log:

```
rasterio._err.CPLE_OpenFailedError: '/vsis3/italy-composite/rasters/italy_2018_red.tif' not recognized as a supported file format.
```

The same tiles worked fine when reloading, it seemed unpredictable, but happened a lot after a while (~10% of all requests).

### Workaround

After a while, we found this issue (#1900) where GDAL would give different error messages depending whether `GDAL_DISABLE_READDIR_ON_OPEN` is set or not.

Since our rasters lie in S3 buckets with > 10,000 files, we use `GDAL_DISABLE_READDIR_ON_OPEN=EMPTY_DIR` by default. We figured that using this flag might obscure the real error (as in #1900). But by switching to `GDAL_DISABLE_READDIR_ON_OPEN=FALSE`, the entire problem went away instead! So we do get the problem when the flag is `TRUE` or `EMPTY_DIR`, but `FALSE` works fine.

This workaround seems to work for the moment, but we do take a small performance hit by doing that.

### Environment

The problem has persisted across different environments for the past 6 months or so. Specifically, we observed it with GDAL 2.2.4, 2.3.2, and 2.4.0 (under Linux).

---

Let me know if we can do anything to help debug this.

**Comment by rouault** [comment]
Collecting logs with CPL_DEBUG=ON and CPL_CURL_VERBOSE=ON enabled might help better identifying the issue.

**Comment by j08lue** [comment]
OK, here are the logs for a failing request, from beginning to end:
```
[1572344645451] [INFO] 2019-10-29T10:24:05.434Z b201b38a-7ddb-46ce-9055-a9f5fa775fa4 <my ip> - - [29/Oct/2019:10:24:05 +0000] "GET /rgb/S2/32UNG/20191027/13/4320/2572.png HTTP/1.1" 500 290 <my hostname> "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:70.0) Gecko/20100101 Firefox/70.0" 0/745.317
[1572344645454] [DEBUG] 2019-10-29T10:24:05.454Z 515f6618-5220-4f4b-bbfd-64f24d2b9d3a streaming subsegments...
[1572344645455] [DEBUG] 2019-10-29T10:24:05.455Z 515f6618-5220-4f4b-bbfd-64f24d2b9d3a sending: {"format":"json","version":1}
{"end_time": 1572344645.4544816, "id": "043a7e287f015666", "in_progress": false, "name": "contrast_stretch", "namespace": "local", "parent_id": "3e99c346e07f6599", "start_time": 1572344645.45284, "trace_id": "1-5db81344-7ed25f3b0dec42fad332c3eb", "type": "subsegment"} to 169.254.79.2:2000.
[1572344645457] [DEBUG] 2019-10-29T10:24:05.456Z ffa48e3b-0d91-490d-94b9-7c3b4047e25d CPLE_None in GDAL: GDALClose(/vsis3/mybucket/S2_32UNG_20191027_B02.tif, this=0x7f097800c820)
[1572344645457] [DEBUG] 2019-10-29T10:24:05.457Z ffa48e3b-0d91-490d-94b9-7c3b4047e25d Exiting env context: <rasterio.env.Env object at 0x7f098be85e48>
[1572344645457] [DEBUG] 2019-10-29T10:24:05.457Z ffa48e3b-0d91-490d-94b9-7c3b4047e25d Cleared existing <rasterio._env.GDALEnv object at 0x7f097d5e6f98> options
[1572344645457] [DEBUG] 2019-10-29T10:24:05.457Z ffa48e3b-0d91-490d-94b9-7c3b4047e25d Stopped GDALEnv <rasterio._env.GDALEnv object at 0x7f097d5e6f98>.
[1572344645457] [DEBUG] 2019-10-29T10:24:05.457Z ffa48e3b-0d91-490d-94b9-7c3b4047e25d Exiting outermost env
[1572344645457] [DEBUG] 2019-10-29T10:24:05.457Z ffa48e3b-0d91-490d-94b9-7c3b4047e25d Exited env context: <rasterio.env.Env object at 0x7f098be85e48>
[1572344645457] [DEBUG] 2019-10-29T10:24:05.457Z ffa48e3b-0d91-490d-94b9-7c3b4047e25d streaming subsegments...
[1572344645457] [DEBUG] 2019-10-29T10:24:05.457Z ffa48e3b-0d91-490d-94b9-7c3b4047e25d sending: {"format":"json","version":1}
{"end_time": 1572344645.4574916, "id": "fb7de3ddd385056f", "in_progress": false, "name": "get_raster_tile", "namespace": "local", "parent_id": "7f70fe305032148a", "start_time": 1572344644.7777433, "trace_id": "1-5db81344-772ce55a4bd2bf003d788600", "type": "subsegment"} to 169.254.79.2:2000.
[1572344645457] [DEBUG] 2019-10-29T10:24:05.439Z 0a83b00e-77dd-432b-8f7a-2f250eaa0db1 CPLE_None in GDAL: GDALWarpKernel()::GWKNearestNoMasksOrDstDensityOnlyShort() Src=6435,4588,284x132 Dst=0,0,285x128
[1572344645457] [DEBUG] 2019-10-29T10:24:05.440Z 0a83b00e-77dd-432b-8f7a-2f250eaa0db1 CPLE_None in GDAL: GDALWarpKernel()::GWKNearestNoMasksOrDstDensityOnlyShort() Src=6437,4714,284x132 Dst=0,128,285x128
[1572344645457] [DEBUG] 2019-10-29T10:24:05.441Z 0a83b00e-77dd-432b-8f7a-2f250eaa0db1 CPLE_None in VSICURL: Downloading 137734662-137925477 (https://mybucket.s3.amazonaws.com/S2_32UNG_20191027_B03.tif)...
[1572344645457] * Couldn't find host mybucket.s3.amazonaws.com in the .netrc file; using defaults
[1572344645457] * Found bundle for host mybucket.s3.amazonaws.com: 0x7fa70408a0a0 [can pipeline]
[1572344645457] * Re-using existing connection! (#1) with host mybucket.s3.amazonaws.com
[1572344645457] * Connected to mybucket.s3.amazonaws.com (52.219.74.145) port 443 (#1)
[1572344645457] > GET /S2_32UNG_20191027_B03.tif HTTP/1.1
[1572344645457] Host: mybucket.s3.amazonaws.com
[1572344645457] Accept: */*
[1572344645457] Range: bytes=137734662-137925477
[1572344645457] x-amz-date: 20191029T102405Z
[1572344645457] x-amz-content-sha256: <removed this>
[1572344645457] X-Amz-Security-Token: <removed this> 
[1572344645457] Authorization: AWS4-HMAC-SHA256 Credential=<removed this>
[1572344645457] < HTTP/1.1 206 Partial Content
[1572344645457] < x-amz-id-2: <removed this>
[1572344645457] < x-amz-request-id: 335E849F34B82A4D
[1572344645457] < Date: Tue, 29 Oct 2019 10:24:06 GMT
[1572344645457] < Last-Modified: Sun, 27 Oct 2019 16:55:52 GMT
[1572344645457] < ETag: "06e26cd4646bac96fa2c36c3f5d21e4a-28"
[1572344645457] < Accept-Ranges: bytes
[1572344645457] < Content-Range: bytes 136934023-137125206/227805393
[1572344645457] < Content-Type: binary/octet-stream
[1572344645457] < Content-Length: 191184
[1572344645457] < Server: AmazonS3
[1572344645457] <
[1572344645458] < HTTP/1.1 206 Partial Content
[1572344645458] < x-amz-id-2: <removed this>
[1572344645458] < x-amz-request-id: DEBA8469664CC88E
[1572344645458] < Date: Tue, 29 Oct 2019 10:24:06 GMT
[1572344645458] < Last-Modified: Sun, 27 Oct 2019 16:55:28 GMT
[1572344645458] < ETag: "8f8ff7b6121ec6a61591aae3e19da1bb-28"
[1572344645458] < Accept-Ranges: bytes
[1572344645458] < Content-Range: bytes 137734662-137925477/228844873
[1572344645458] < Content-Type: binary/octet-stream
[1572344645458] < Content-Length: 190816
[1572344645464] [DEBUG] 2019-10-29T10:24:05.446Z 1076ab39-d7ae-4d34-8321-4ce34331e7cf CPLE_None in GDAL: GDALWarpKernel()::GWKNearestNoMasksOrDstDensityOnlyShort() Src=6451,5672,283x34 Dst=0,256,285x29
[1572344645464] * SSL connection using TLSv1.2 / ECDHE-RSA-AES128-SHA
[1572344645464] * ALPN, server did not agree to a protocol
[1572344645464] * Server certificate:
[1572344645464] *  subject: C=US; ST=Washington; L=Seattle; O=Amazon.com Inc.; CN=*.s3.amazonaws.com
[1572344645464] *  start date: Nov  7 00:00:00 2018 GMT
[1572344645464] *  expire date: Feb  7 12:00:00 2020 GMT
[1572344645464] *  subjectAltName: host "mybucket.s3.amazonaws.com" matched cert's "*.s3.amazonaws.com"
[1572344645464] *  issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=DigiCert Baltimore CA-2 G2
[1572344645464] *  SSL certificate verify ok.
[1572344645464] > GET /S2_32UNG_20191027_B02.tif HTTP/1.1
[1572344645464] Host: mybucket.s3.amazonaws.com
[1572344645464] Accept: */*
[1572344645464] Range: bytes=145794048-145988815
[1572344645464] x-amz-date: 20191029T102405Z
[1572344645464] x-amz-content-sha256: <removed this> 
[1572344645464] X-Amz-Security-Token: <removed this>
[1572344645464] Authorization: AWS4-HMAC-SHA256 Credential=<removed this>
[1572344645467] [DEBUG] 2019-10-29T10:24:05.466Z 1076ab39-d7ae-4d34-8321-4ce34331e7cf Output nodata value read from file: None
[1572344645467] [DEBUG] 2019-10-29T10:24:05.467Z 1076ab39-d7ae-4d34-8321-4ce34331e7cf Output nodata values: [None]
[1572344645467] [DEBUG] 2019-10-29T10:24:05.467Z 1076ab39-d7ae-4d34-8321-4ce34331e7cf all_valid: False
[1572344645467] [DEBUG] 2019-10-29T10:24:05.467Z 1076ab39-d7ae-4d34-8321-4ce34331e7cf mask_flags: ([<MaskFlags.per_dataset: 2>, <MaskFlags.alpha: 4>], [<MaskFlags.all_valid: 1>])
[1572344645467] [DEBUG] 2019-10-29T10:24:05.467Z 1076ab39-d7ae-4d34-8321-4ce34331e7cf Jump straight to _read()
[1572344645467] [DEBUG] 2019-10-29T10:24:05.467Z 1076ab39-d7ae-4d34-8321-4ce34331e7cf Window: Window(col_off=2, row_off=2, width=281, height=281)
[1572344645467] [DEBUG] 2019-10-29T10:24:05.467Z 1076ab39-d7ae-4d34-8321-4ce34331e7cf IO window xoff=2.0 yoff=2.0 width=281.0 height=281.0
[1572344645471] [DEBUG] 2019-10-29T10:24:05.471Z 515f6618-5220-4f4b-bbfd-64f24d2b9d3a streaming subsegments...
[1572344645472] [DEBUG] 2019-10-29T10:24:05.471Z 515f6618-5220-4f4b-bbfd-64f24d2b9d3a sending: {"format":"json","version":1}
{"end_time": 1572344645.4714408, "id": "80df56f504633396", "in_progress": false, "name": "get_metadata", "namespace": "local", "parent_id": "3e99c346e07f6599", "start_time": 1572344645.4709547, "trace_id": "1-5db81344-7ed25f3b0dec42fad332c3eb", "type": "subsegment"} to 169.254.79.2:2000.
[1572344645472] [DEBUG] 2019-10-29T10:24:05.472Z 515f6618-5220-4f4b-bbfd-64f24d2b9d3a streaming subsegments...
[1572344645472] [DEBUG] 2019-10-29T10:24:05.472Z 515f6618-5220-4f4b-bbfd-64f24d2b9d3a sending: {"format":"json","version":1}
{"cause": "9ad5293588eb8da3", "end_time": 1572344645.4724696, "fault": true, "id": "3e99c346e07f6599", "in_progress": false, "name": "rgb_handler", "namespace": "local", "parent_id": "0a876b9a4ddd46f8", "start_time": 1572344644.6868117, "trace_id": "1-5db81344-7ed25f3b0dec42fad332c3eb", "type": "subsegment"} to 169.254.79.2:2000.
[1572344645473] [ERROR] 2019-10-29T10:24:05.473Z 515f6618-5220-4f4b-bbfd-64f24d2b9d3a Exception on /rgb/S2/32UNG/20191027/13/4325/2572.png [GET]
Traceback (most recent call last):
  File "rasterio/_base.pyx", line 216, in rasterio._base.DatasetBase.__init__
  File "rasterio/_shim.pyx", line 66, in rasterio._shim.open_dataset
  File "rasterio/_err.pyx", line 205, in rasterio._err.exc_wrap_pointer
rasterio._err.CPLE_OpenFailedError: '/vsis3/mybucket/S2_32UNG_20191027_B02.tif' not recognized as a supported file format.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/task/terracotta/drivers/raster_base.py", line 470, in _get_raster_tile
  src = es.enter_context(rasterio.open(path))
  File "/var/task/rasterio/env.py", line 445, in wrapper
  return f(*args, **kwds)
  File "/var/task/rasterio/__init__.py", line 219, in open
  s = DatasetReader(path, driver=driver, sharing=sharing, **kwargs)
  File "rasterio/_base.pyx", line 218, in rasterio._base.DatasetBase.__init__
rasterio.errors.RasterioIOError: '/vsis3/mybucket/S2_32UNG_20191027_B02.tif' not recognized as a supported file format.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/task/flask/app.py", line 2311, in wsgi_app
  response = self.full_dispatch_request()
  File "/var/task/flask/app.py", line 1834, in full_dispatch_request
  rv = self.handle_user_exception(e)
  File "/var/task/flask/app.py", line 1737, in handle_user_exception
  reraise(exc_type, exc_value, tb)
  File "/var/task/flask/_compat.py", line 36, in reraise
  raise value
  File "/var/task/flask/app.py", line 1832, in full_dispatch_request
  rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1818, in dispatch_request
  return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/task/terracotta/server/flask_api.py", line 50, in inner
  return fun(*args, **kwargs)
  File "/var/task/terracotta/server/rgb.py", line 85, in get_rgb
  return _get_rgb_image(keys, tile_xyz=tile_xyz)
  File "/var/task/terracotta/server/rgb.py", line 132, in _get_rgb_image
  some_keys, rgb_values, stretch_ranges=stretch_ranges, tile_xyz=tile_xyz, **options
  File "/var/lang/lib/python3.6/contextlib.py", line 52, in inner
  return func(*args, **kwds)
  File "/var/task/terracotta/handlers/rgb.py", line 85, in rgb
  band_data = band_data_future.result()
  File "/var/lang/lib/python3.6/concurrent/futures/_base.py", line 425, in result
  return self.__get_result()
  File "/var/lang/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
  raise self._exception
  File "/var/lang/lib/python3.6/concurrent/futures/thread.py", line 56, in run
  result = self.fn(*self.args, **self.kwargs)
  File "/var/task/cachetools/__init__.py", line 87, in wrapper
  v = method(self, *args, **kwargs)
  File "/var/lang/lib/python3.6/contextlib.py", line 52, in inner
  return func(*args, **kwds)
  File "/var/task/terracotta/drivers/raster_base.py", line 472, in _get_raster_tile
  raise IOError('error while reading file {}'.format(path))
OSError: error while reading file s3://mybucket/S2_32UNG_20191027_B02.tif
```

**Comment by dionhaefner** [comment]
I think `streaming subsegments` is indeed from X-Ray and shouldn't have anything to do with this. It's basically just logging data getting sent after certain functions exit (and post-mortem).

**Comment by rouault** [comment]
I don't see anything obviously wrong in the above log. It gets a valid HTTP 206 response. Is there a possibillity that your code is multi-threaded ? If so, there's a known issue with /vsicurl/ (or /vsis3/) caching with multithreading

**Comment by rcoup** [comment]
> I don't see anything obviously wrong in the above log. It gets a valid HTTP 206 response. 

The first request does, but the second (for bytes 145794048-145988815) doesn't have a response logged — does that imply it's a TCP (rather than HTTP) error?

**Comment by rouault** [comment]
> The first request does, but the second (for bytes 145794048-145988815) doesn't have a response logged — does that imply it's a TCP (rather than HTTP) error?

Indeed. But even if that was the case, there should be a VSICURL trace "Got response_code=" (from https://github.com/OSGeo/gdal/blob/master/gdal/port/cpl_vsil_curl.cpp#L1333) , which I can't see . Isn't there something filtering out debug traces ?

**Comment by dionhaefner** [comment]
The code *is* multi-threaded, but it only reads different files in parallel (here `_B03.tif` and `_B02.tif`), never the same file. Is that still a problem?

**Comment by rouault** [comment]
>  Is that still a problem?

yes, that's exactly the scenario of https://github.com/OSGeo/gdal/issues/1244 AFAICS. To confirm this is really the same issue, can you add some locking around GDAL calls / disable multi-threading ?

**Comment by dionhaefner** [comment]
Hmm, good point. I'll poke around a bit and let you know.

**Comment by dionhaefner** [comment]
I can confirm that this doesn't seem to occur without multi-threading.

**Comment by rouault** [comment]
OK, so closing this as a duplicate of #1244
